### PR TITLE
created integration test that asserts cursor functionality

### PR DESCRIPTION
**Pull request by EasterTheBunny**

### CODE REVIEW

**Review by silaslenihan — DISMISSED** [review]

**Review by vyzaldysanchez — DISMISSED** [review]

**Review by silaslenihan — APPROVED** [review]

File: pkg/types/interfacetests/chain_components_interface_tests.go [hunk 1]
```diff
@@ -794,6 +794,57 @@ func runQueryKeyInterfaceTests[T TestingT[T]](t T, tester ChainComponentsInterfa
...
+						}
+
+						limit.Limit = query.CursorLimit(sequences[len(sequences)-1].Cursor, query.CursorFollowing, 2)
+						allSequences = append(allSequences, sequences...)
+					}
+
+					return len(allSequences) == len(testStructs) &&
+						reflect.DeepEqual(&testStructs[0], allSequences[0].Data) &&
+						reflect.DeepEqual(&testStructs[len(testStructs)-1], allSequences[len(testStructs)-1].Data)
+				}, tester.MaxWaitTimeForEvents(), 50*time.Millisecond)
```
Comment (jmank88) [inline pkg/types/interfacetests/chain_components_interface_tests.go:845]: There's no need to check this often, especially with such complex code:
```suggestion
				}, tester.MaxWaitTimeForEvents(), 500*time.Millisecond)
```

File: pkg/types/interfacetests/chain_components_interface_tests.go [hunk 2]
```diff
@@ -794,6 +794,57 @@ func runQueryKeyInterfaceTests[T TestingT[T]](t T, tester ChainComponentsInterfa
 				}, tester.MaxWaitTimeForEvents(), time.Millisecond*10)
```
Comment (jmank88) [inline pkg/types/interfacetests/chain_components_interface_tests.go:794]: This is out of scope, but 10ms seems way too fast:
```suggestion
				}, tester.MaxWaitTimeForEvents(), time.Millisecond*500)
```
There is typically little payoff to doing these checks so fast. You risk keeping the CPU so busy that it slows down or cannot even make progress on the test, and the best case payoff is only to finish earlier by the max difference in the durations (assuming parallel tests).

**Review by jmank88 — DISMISSED** [review]

File: pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go [hunk 3]
```diff
@@ -621,21 +622,39 @@ func (f *fakeContractReader) QueryKey(_ context.Context, bc types.BoundContract,
...
+			}
+
+			if (len(filter.Expressions) == 0 || doAppend) && !skipAppend {
 				if isValueType {
 					value, err := values.Wrap(trigger.testStruct)
 					if err != nil {
 						return nil, err
 					}
-					sequences = append(sequences, types.Sequence{Data: &value})
+					sequences = append(sequences, types.Sequence{Cursor: fmt.Sprintf("%d", idx), Data: &value})
```
Comment (jmank88) [inline pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go:645]: Could shorten these a bit?
```suggestion
					sequences = append(sequences, types.Sequence{Cursor: strconv.Itoa(idx), Data: &value})
```

**Review by silaslenihan — APPROVED** [review]

**Review by jmank88 — APPROVED** [review]

**Review by Farber98 — APPROVED** [review]

**Review by ilija42 — COMMENTED** [review]
Looks good, but there are failing tests

File: pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go [hunk 4]
```diff
@@ -337,7 +337,7 @@ func (it *fakeContractReaderInterfaceTester) GenerateBlocksTillConfidenceLevel(t
 }
 
 func (it *fakeContractReaderInterfaceTester) MaxWaitTimeForEvents() time.Duration {
-	return time.Millisecond * 100
+	return time.Second * 10
```
Comment (ilija42) [inline pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go:340]: why, this is a 100x increase?

Comment (ilija42) [reply pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go]: Approved, but we should look into optimising these. Which test was this?

**Review by ilija42 — APPROVED** [review]

File: pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go (hunk 4 above)
Comment (silaslenihan) [reply pkg/loop/internal/relayer/pluginprovider/contractreader/contract_reader_test.go]: Any test that uses `MaxWaitTimeForEvents()` was failing in this PR before I made this change

**Review by silaslenihan — APPROVED** [review]

**Pull request merged**